File: Atividade 002 (Dawkins' Weasel Program)/WeaselProgram.c
#include<stdio.h>
#include<stdlib.h>
#include<time.h>

char alphabet[28] = {'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J',
                    'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R', 'S', 'T',
                    'U', 'V', 'W', 'X', 'Y', 'Z', ' '};

void print_text(char *text, int seq){
    printf("Sequence %d: ", seq);
    for(int i = 0; i < 28; i++){
        printf("%c", text[i]);
    }
    printf("\n");
}

int main(){
    srand(time(NULL));
    char text_target[28] = "METHINKS IT IS LIKE A WEASEL";
    char sequence[28];
    char copies[100][28];
    int best_pontuation = -1, conting_generations = 0, pontuation;
    for(int i = 0; i < 28; i++){
        sequence[i] = alphabet[rand()%28];
    }
    while(best_pontuation < 28){
        conting_generations++;
        for(int i = 0; i < 100; i++){
            for(int j = 0; j < 28; j++){
                copies[i][j] = sequence[j];
            }
        }
        for(int i = 0; i < 100; i++){
            for(int j = 0; j < 28; j++){
                if(rand()%100 < 5){
                    copies[i][j] = alphabet[rand()%28];
                }
            }
        }
        for(int i = 0; i < 100; i++){
            pontuation = 0;
            for(int j = 0; j < 28; j++){
                if(copies[i][j] == text_target[j]){
                    pontuation++;
                }
            }
            if(pontuation > best_pontuation){
                best_pontuation = pontuation;
                for(int k = 0; k < 28; k++){
                    sequence[k] = copies[i][k];
                }
            }
        }
        print_text(sequence, conting_generations);
    }
    printf("Generations: %d\n", conting_generations);
    return 0;
}

File: Atividade 002 (Dawkins' Weasel Program)/WeaselProgram.py
import random


char = ['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J',
        'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R', 'S', 'T',
        'U', 'V', 'W', 'X', 'Y', 'Z', ' ']


def _text_init():
    text = ""
    cont = 0
    while(cont < 28):
        text += random.choice(char)
        cont += 1
    return text

conting_generations = 0
text_target = "METHINKS IT IS LIKE A WEASEL"
sequence = _text_init()
best_pontuation = -1
while(best_pontuation < 28):
    conting_generations += 1
    copies = [sequence]*100
    for i in range(100):
        text_help = ""
        for letter in copies[i]:
            if(random.random() <= 0.05):
                text_help += random.choice(char)
            else:
                text_help += letter
        copies[i] = text_help
    for text in copies:
        pontuation = 0
        for i in range(28):
            if(text[i] == text_target[i]):
                pontuation += 1
        if(pontuation > best_pontuation):
            best_pontuation = pontuation
            sequence = text
    print("sequence %d: %s" % (conting_generations, sequence))
print("Generations:", conting_generations)